### DISCUSSION TITLE
Driver crashes System after add new devices

### DISCUSSION THREAD

**Discussion by stepi2** [Louisvdw/dbus-serialbattery#210] [Q&A] [answered]
First of all, a big thank you for the great work, especially on Louis.

The driver has worked very well for the past few months.

Last week I added two more devices (smartsolar + second energymeter).

The BMS is displayed in the remote console for a few seconds, then "not connected" for a few seconds.
If "not connected" is displayed, the ESS no longer works (Multiplus goes down to stop discharging).
The remote console is very slow and lagged.

To me it seems to be a performance issue. The same problem occurs with daly and JK-BMS

System: Raspberry 4 with venus OS
2x rs485 energy meters
1x multi plus II
1x smart solar
1x modbus tcp (SMA)
1xmqtt server enabled

Any tips?

Is it possible to use the driver only as a monitor without controlling the discharge current?
You can switch off CCL and CVL

Thanks Stefan

logfile:
`@400000006320d11b2ab8e5b4 INFO:SerialBattery:Starting dbus-serialbattery
@400000006320d11b2ac3691c INFO:SerialBattery:dbus-serialbattery v0.12b3
@400000006320d11b2ae39b4c INFO:SerialBattery:Testing LltJbd
@400000006320d11b3b740a14 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d11b3b898de4 INFO:SerialBattery:Testing Ant
@400000006320d11c1037b1fc ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d11c104d0304 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006320d11c10556774 INFO:SerialBattery:Testing Daly
@400000006320d11c207d0e04 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d11c20920d04 INFO:SerialBattery:Testing Daly
@400000006320d11c30b951ec ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d11c30cd7244 INFO:SerialBattery:Testing Jkbms
@400000006320d11c38190bbc INFO:SerialBattery:Connection established to Jkbms
@400000006320d11c3875ed14 INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB3
@400000006320d11c38760c54 INFO:SerialBattery:=== Settings ===
@400000006320d11c38761424 INFO:SerialBattery:> Connection voltage 52.82V | current -0.0A | SOC 31%
@400000006320d11c387627ac INFO:SerialBattery:> Cell count 16 | cells populated 0
@400000006320d11c38763364 INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@400000006320d11c38764304 INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.9V | MAX_CELL_VOLTAGE 3.45V
@400000006320d11c3a550584 INFO:SerialBattery:DeviceInstance = 4
@400000006320d11c3a5d8164 INFO:SerialBattery:com.victronenergy.battery.ttyUSB3
@400000006320d153060e50dc Traceback (most recent call last):
@400000006320d153060e684c   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 163, in publish_battery
@400000006320d153060e7fbc     success = self.battery.refresh_data()
@400000006320d153060e8b74   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 53, in refresh_data
@400000006320d153060e9b14     result = self.read_status_data()
@400000006320d153060f1044   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 66, in read_status_data
@400000006320d153061190e4     status_data = self.read_serial_data_jkbms(self.command_status)
@400000006320d1530611a46c   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 150, in read_serial_data_jkbms
@400000006320d1530611bbdc     data = read_serial_data(command, self.port, self.baud_rate, self.LENGTH_POS, self.LENGTH_CHECK,None, self.LENGTH_SIZE)
@400000006320d1530611d34c   File "/opt/victronenergy/dbus-serialbattery/utils.py", line 77, in read_serial_data
@400000006320d153061359ec     with serial.Serial(port, baudrate=baud, timeout=0.1) as ser:
@400000006320d15306136d74   File "/usr/lib/python3.8/site-packages/serial/serialutil.py", line 240, in __init__
@400000006320d153061380fc     self.open()
@400000006320d153061388cc   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 288, in open
@400000006320d1530613986c     self._update_rts_state()
@400000006320d1530613f62c   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 627, in _update_rts_state
@400000006320d1530614f7fc     fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_RTS_str)
@400000006320d1530615079c OSError: [Errno 5] Input/output error
@400000006320d1752f11c824 INFO:SerialBattery:Starting dbus-serialbattery
@400000006320d1752f199054 INFO:SerialBattery:dbus-serialbattery v0.12b3
@400000006320d1752f39225c INFO:SerialBattery:Testing LltJbd
@400000006320d17603e8a9ec ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d17603fce59c INFO:SerialBattery:Testing Ant
@400000006320d1761411b7dc ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d17614852e74 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006320d176148c5e4c INFO:SerialBattery:Testing Daly
@400000006320d17624b0cca4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d17624c61dac INFO:SerialBattery:Testing Daly
@400000006320d17635282974 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d176353ebacc INFO:SerialBattery:Testing Jkbms
@400000006320d176383d58b4 INFO:SerialBattery:Connection established to Jkbms
@400000006320d176383d77f4 INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB3
@400000006320d1763845346c INFO:SerialBattery:=== Settings ===
@400000006320d176384fe2cc INFO:SerialBattery:> Connection voltage 52.82V | current -0.0A | SOC 31%
@400000006320d176385a818c INFO:SerialBattery:> Cell count 16 | cells populated 0
@400000006320d176386314dc INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@400000006320d176386c0204 INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.9V | MAX_CELL_VOLTAGE 3.45V
@400000006320d1763aee1134 INFO:SerialBattery:DeviceInstance = 4
@400000006320d1763afb5b8c INFO:SerialBattery:com.victronenergy.battery.ttyUSB3
@400000006320d18a047067ec Traceback (most recent call last):
@400000006320d18a04707f5c   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 163, in publish_battery
@400000006320d18a047092e4     success = self.battery.refresh_data()
@400000006320d18a04709e9c   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 53, in refresh_data
@400000006320d18a0470b224     result = self.read_status_data()
@400000006320d18a0470c5ac   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 66, in read_status_data
@400000006320d18a04734264     status_data = self.read_serial_data_jkbms(self.command_status)
@400000006320d18a047355ec   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 150, in read_serial_data_jkbms
@400000006320d18a04736974     data = read_serial_data(command, self.port, self.baud_rate, self.LENGTH_POS, self.LENGTH_CHECK,None, self.LENGTH_SIZE)
@400000006320d18a047384cc   File "/opt/victronenergy/dbus-serialbattery/utils.py", line 77, in read_serial_data
@400000006320d18a04740d84     with serial.Serial(port, baudrate=baud, timeout=0.1) as ser:
@400000006320d18a04741d24   File "/usr/lib/python3.8/site-packages/serial/serialutil.py", line 240, in __init__
@400000006320d18a047430ac     self.open()
@400000006320d18a0474387c   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 286, in open
@400000006320d18a04744c04     self._update_dtr_state()
@400000006320d18a0475bf1c   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 634, in _update_dtr_state
@400000006320d18a04767e84     fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_DTR_str)
@400000006320d18a047699dc OSError: [Errno 5] Input/output error
@400000006320d1ab02e631f4 INFO:SerialBattery:Starting dbus-serialbattery
@400000006320d1ab02f3071c INFO:SerialBattery:dbus-serialbattery v0.12b3
@400000006320d1ab033a5504 INFO:SerialBattery:Testing LltJbd
@400000006320d1ab136be6a4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1ab1386311c INFO:SerialBattery:Testing Ant
@400000006320d1ab23be8d04 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1ab23d97f74 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006320d1ab23d9a684 INFO:SerialBattery:Testing Daly
@400000006320d1ab34828b04 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1ab34b3a93c INFO:SerialBattery:Testing Daly
@400000006320d1ac098a11ec ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1ac099ee20c INFO:SerialBattery:Testing Jkbms
@400000006320d1ac0c77eec4 INFO:SerialBattery:Connection established to Jkbms
@400000006320d1ac0c8085fc INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB3
@400000006320d1ac0c8734ec INFO:SerialBattery:=== Settings ===
@400000006320d1ac0c8fbc84 INFO:SerialBattery:> Connection voltage 52.82V | current -0.0A | SOC 31%
@400000006320d1ac0c96966c INFO:SerialBattery:> Cell count 16 | cells populated 0
@400000006320d1ac0c9d4944 INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@400000006320d1ac0ca4232c INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.9V | MAX_CELL_VOLTAGE 3.45V
@400000006320d1ac0fb3090c INFO:SerialBattery:DeviceInstance = 4
@400000006320d1ac0fb3207c INFO:SerialBattery:com.victronenergy.battery.ttyUSB3
@400000006320d1c013d8b61c Traceback (most recent call last):
@400000006320d1c013d8d174   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 163, in publish_battery
@400000006320d1c013d8e4fc     success = self.battery.refresh_data()
@400000006320d1c013d8f49c   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 53, in refresh_data
@400000006320d1c013d9043c     result = self.read_status_data()
@400000006320d1c013d90ff4   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 66, in read_status_data
@400000006320d1c013ddd69c     status_data = self.read_serial_data_jkbms(self.command_status)
@400000006320d1c013ddee0c   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 150, in read_serial_data_jkbms
@400000006320d1c013de057c     data = read_serial_data(command, self.port, self.baud_rate, self.LENGTH_POS, self.LENGTH_CHECK,None, self.LENGTH_SIZE)
@400000006320d1c013de20d4   File "/opt/victronenergy/dbus-serialbattery/utils.py", line 77, in read_serial_data
@400000006320d1c013ded484     with serial.Serial(port, baudrate=baud, timeout=0.1) as ser:
@400000006320d1c013df5954   File "/usr/lib/python3.8/site-packages/serial/serialutil.py", line 240, in __init__
@400000006320d1c013df6cdc     self.open()
@400000006320d1c013df74ac   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 288, in open
@400000006320d1c013df8834     self._update_rts_state()
@400000006320d1c013e00534   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 627, in _update_rts_state
@400000006320d1c013e358c4     fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_RTS_str)
@400000006320d1c013e3647c OSError: [Errno 5] Input/output error
@400000006320d1e22325db2c INFO:SerialBattery:Starting dbus-serialbattery
@400000006320d1e2232e37cc INFO:SerialBattery:dbus-serialbattery v0.12b3
@400000006320d1e2234e5e44 INFO:SerialBattery:Testing LltJbd
@400000006320d1e2336c6bcc ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e233808c24 INFO:SerialBattery:Testing Ant
@400000006320d1e308102004 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e308246384 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006320d1e3082bcdf4 INFO:SerialBattery:Testing Daly
@400000006320d1e3187b3064 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e318935fcc INFO:SerialBattery:Testing Daly
@400000006320d1e329141b0c ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e3292b0e0c INFO:SerialBattery:Testing Jkbms
@400000006320d1e3398bf4dc ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e3399f5d9c INFO:SerialBattery:Testing Sinowealth
@400000006320d1e40e71fd8c ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e40ece9c7c INFO:SerialBattery:Testing Renogy
@400000006320d1e41f10e374 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e41f30db0c INFO:SerialBattery:Testing Revov
@400000006320d1e42f89f9ac ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e42fa3d6c4 ERROR:SerialBattery:read_serial_data_revov::Serial Data is Bad
@400000006320d1e42faeb01c INFO:SerialBattery:Testing Ecs
@400000006320d1e51584fad4 INFO:SerialBattery:Testing LltJbd
@400000006320d1e525c6b52c ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e525d2849c INFO:SerialBattery:Testing Ant
@400000006320d1e5363697ec ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e53640bd94 ERROR:SerialBattery:>>> ERROR: Incorrect Data
@400000006320d1e53647ce2c INFO:SerialBattery:Testing Daly
@400000006320d1e60ae05704 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e60aecc2b4 INFO:SerialBattery:Testing Daly
@400000006320d1e61af976d4 ERROR:SerialBattery:>>> ERROR: No reply - returning
@400000006320d1e61b07b744 INFO:SerialBattery:Testing Jkbms
@400000006320d1e61dc8a04c INFO:SerialBattery:Connection established to Jkbms
@400000006320d1e61dd0ed4c INFO:SerialBattery:Battery connected to dbus from /dev/ttyUSB3
@400000006320d1e61dd7d2ec INFO:SerialBattery:=== Settings ===
@400000006320d1e61ddfe554 INFO:SerialBattery:> Connection voltage 52.82V | current -0.0A | SOC 31%
@400000006320d1e61de70d5c INFO:SerialBattery:> Cell count 16 | cells populated 0
@400000006320d1e61dee6c14 INFO:SerialBattery:> CCL Charge NoneA | DCL Discharge NoneA
@400000006320d1e61df65b54 INFO:SerialBattery:> MIN_CELL_VOLTAGE 2.9V | MAX_CELL_VOLTAGE 3.45V
@400000006320d1e62066365c INFO:SerialBattery:DeviceInstance = 4
@400000006320d1e62070cd4c INFO:SerialBattery:com.victronenergy.battery.ttyUSB3
@400000006320d1f524af1b0c Traceback (most recent call last):
@400000006320d1f524af327c   File "/opt/victronenergy/dbus-serialbattery/dbushelper.py", line 163, in publish_battery
@400000006320d1f524af4604     success = self.battery.refresh_data()
@400000006320d1f524af55a4   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 53, in refresh_data
@400000006320d1f524af6544     result = self.read_status_data()
@400000006320d1f524af70fc   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 66, in read_status_data
@400000006320d1f524b25efc     status_data = self.read_serial_data_jkbms(self.command_status)
@400000006320d1f524b27284   File "/opt/victronenergy/dbus-serialbattery/jkbms.py", line 150, in read_serial_data_jkbms
@400000006320d1f524b2860c     data = read_serial_data(command, self.port, self.baud_rate, self.LENGTH_POS, self.LENGTH_CHECK,None, self.LENGTH_SIZE)
@400000006320d1f524b2a164   File "/opt/victronenergy/dbus-serialbattery/utils.py", line 77, in read_serial_data
@400000006320d1f524b32a1c     with serial.Serial(port, baudrate=baud, timeout=0.1) as ser:
@400000006320d1f524b339bc   File "/usr/lib/python3.8/site-packages/serial/serialutil.py", line 240, in __init__
@400000006320d1f524b34d44     self.open()
@400000006320d1f524b35514   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 288, in open
@400000006320d1f524b3689c     self._update_rts_state()
@400000006320d1f524b400f4   File "/usr/lib/python3.8/site-packages/serial/serialposix.py", line 627, in _update_rts_state
@400000006320d1f524b4c444     fcntl.ioctl(self.fd, TIOCMBIS, TIOCM_RTS_str)
@400000006320d1f524b4d3e4 OSError: [Errno 5] Input/output error
@400000006320d2152f6babdc *** CCGX booted (0) ***`

**Comment by Louisvdw** [comment]
In the log file the one thing that looks funny is this part that show that the charge/discharge current is not set, which cannot be.
> CCL Charge NoneA | DCL Discharge NoneA

Best option would be to try and reinstall the driver.

You must have CVL and CVL, but those can be edit as you need. See the install instructions.

**Comment by stepi2** [comment, marked as answer]
i found the solution...

I used the same physical USB-port on Rapsi for BMS and Victron Smartsolar-charger with an USB-Hub. In  this configuration the  whole venus-system crashes after 20-30 seconds. The ESS-control, mqtt-server an so on goes down and come back after 30-40 minutes and same begins again.

If I use different physical USB-ports all working fine.

I Think there ist a problem in implementation  of the serial-driver.

Now I have all the other devices on the USB-hub and the BMS has his own port...This workaound is OK for me

**Comment by Louisvdw** [comment]
My own setup has MPPTs with a BMS on a HUB to one port on the Pi and another BMS on the Pi directly. 
So I expect it might be something else that caused your issue.
Glad you have it running now.
